File: LeetCode/Medium/0207-course-schedule/0207-course-schedule.java
class Solution {
    ArrayList<ArrayList<Integer>> req;
    HashSet<Integer> visited;
    
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        req = new ArrayList();
        visited = new HashSet();
        
        for (int i = 0; i < numCourses; ++i) req.add(new ArrayList());
        
        for (int i = 0; i < prerequisites.length; ++i)
            req.get(prerequisites[i][0]).add(prerequisites[i][1]);
        
        for (int i = 0; i < numCourses; ++i) 
            if (!dfs(i)) return false;
        
        return true;
    }
    
    private boolean dfs(int course){
        if (visited.contains(course)) return false;
        if (req.get(course).size() == 0) return true;
        
        visited.add(course);
        
        for (int i = 0; i < req.get(course).size(); ++i){
            if (!dfs(req.get(course).get(i))) return false;
        }
        
        visited.remove(course);
        req.set(course, new ArrayList());
        
        return true;
    }
}

File: LeetCode/Medium/0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        req = [[] for i in range(numCourses)]

        for v0, v1 in prerequisites:
            req[v0].append(v1)
        
        visited = set()

        def dfs(course):
            if course in visited:
                return False
            if not req[course]:
                return True
            
            visited.add(course)

            for next in req[course]:
                if not dfs(next):
                    return False
            
            visited.remove(course)
            req[course] = []
            
            return True

        for i in range(numCourses):
            if not dfs(i):
                return False

        return True